osdep/semaphore_osx.c

```diff
@@ -67,32 +69,18 @@ int mp_sem_trywait(mp_sem_t *sem)
     return r;
 }
 
-int mp_sem_timedwait(mp_sem_t *sem, const struct timespec *abs_timeout)
+int mp_sem_timedwait(mp_sem_t *sem, int64_t until)
 {
     while (1) {
         if (!mp_sem_trywait(sem))
             return 0;
 
-        int timeout_ms = -1;
-        if (abs_timeout) {
-            timeout_ms = 0;
-
-            // OSX does not provide clock_gettime() either.
-            struct timeval tv;
-            gettimeofday(&tv, NULL);
-
-            if (abs_timeout->tv_sec >= tv.tv_sec) {
-                long long msec = (abs_timeout->tv_sec - tv.tv_sec) * 1000LL +
-                    abs_timeout->tv_nsec / 1000LL / 1000LL - tv.tv_usec / 1000LL;
-                if (msec > INT_MAX)
-                    msec = INT_MAX;
-                if (msec < 0)
-                    msec = 0;
-                timeout_ms = msec;
-            }
-        }
+        int64_t timeout = MPCLAMP(until - mp_time_ns(),
+                                  MP_TIME_MS_TO_NS(INT_MIN),
+                                  MP_TIME_MS_TO_NS(INT_MAX));
+        timeout /= MP_TIME_MS_TO_NS(1);
```

Comment:
Also, I think it'd be more accurate to do the clamping *after* the division.

Reply:
But in any case, I think it's more natural to do it like this. And IMO more straight=forward to read as well.

```c
int64_t timeout = until - mp_time_ns();
timeout /= MP_TIME_MS_TO_NS(1);
timeout = MPCLAMP(timeout, INT_MIN, INT_MAX);
```

Maybe you can inline the division though.

Reply:
changed, see now